File: gridlock/solver.py
from gridlock.pieces import PIECES
from gridlock.board import Board

GIVEN_SOLUTIONS = [
    'GJhCiDbEAKF',   # Ships with the game
    'JIdKehFgACB',   # Printed on the case
]

total_sols = 0

def check_for_win(board):
    global total_sols
    # Check if all pieces are on the board
    if board.grid.count(46) == 0: # ord('.')
        total_sols += 1
        # print("Found solution:")
        # board.print()    

def solve(board):
    global total_sols
    total_sols = 0
    _solve_board(board)
    return total_sols

def _solve_board(board):    
    # Find the largest piece that isn't on the given board
    fnd = None
    for p in PIECES.values():
        if p.letter not in board.grid and (p.letter | 32) not in board.grid:        
            fnd = p
            break
    if fnd is None:
        # All pieces are on the board        
        return
    
    # Slide and rotate the piece over the board

    for rot in [False, True]:
        if rot and fnd.letter in [65, 70, 74]: # A, F, J
            # Only one orientation for these pieces
            continue
        for y in range(8):
            for x in range(8):
                if board.place_piece(fnd, x, y, rot):
                    if fnd.letter == ord('K'):
                        print("Placed K at", x, y, "rotated?", rot)
                    check_for_win(board)
                    _solve_board(board)
                board.remove_piece(fnd)

if __name__ == "__main__":
    import datetime
    print(datetime.datetime.now())
    b = Board()
    print("Solving...")
    total = solve(b)
    print(f"Total solutions found: {total}")
    print(datetime.datetime.now())

File: pkg/gridlock/solver.go
package gridlock

var TotalCount int
var JustOneSolution Board

func CountSolve(brd *Board, allowed_pieces *[]Piece) (int, Board) {
	TotalCount = 0
	JustOneSolution = CreateBoard(brd.Width, brd.Height)
	countSolutionsRecursive(brd, allowed_pieces)
	return TotalCount, JustOneSolution
}

func countSolutionsRecursive(brd *Board, allowed_pieces *[]Piece) {
	// if TotalCount > 50_000_000 {
	// 	return
	// }
	// Find the largest piece that isn't on the given board
	fnd := -1
	for i, p := range *allowed_pieces {
		if !brd.HasPiece(p.Name) {
			fnd = i
			break
		}
	}
	if fnd < 0 {
		// All pieces are on the board
		TotalCount++
		if TotalCount == 1 {
			copy(JustOneSolution.Cells, brd.Cells)
		}
		for _, p := range (*brd).Cells {
			if p == '.' {
				panic("Invalid solution: empty space found")
			}
		}
		return
	}
	piece := (*allowed_pieces)[fnd]

	// Slide and rotate the piece over the board

	for rot := range 2 {
		if rot == 1 && (piece.Width == piece.Height) {
			// Square pieces are the same no matter the rotation. We
			// can save time by skipping rotations on a square piece.
			continue
		}
		for y := range brd.Height {
			for x := range brd.Width {
				if brd.PlacePiece(piece, x, y, rot == 1) {
					countSolutionsRecursive(brd, allowed_pieces)
				}
				brd.RemovePiece(piece)
			}
		}
	}
}

// Place all remaining pieces on the board. Add any solutions found to the
// given list of solutions.
func Solve(brd *Board, allowed_pieces *[]Piece, retSolutions *[]Board) {

	// Find the largest piece that isn't on the given board
	fnd := -1
	for i, p := range *allowed_pieces {
		if !brd.HasPiece(p.Name) {
			fnd = i
			break
		}
	}
	if fnd < 0 {
		// All pieces are on the board
		nb := CreateBoard(brd.Width, brd.Height)
		copy(nb.Cells, brd.Cells)
		*retSolutions = append(*retSolutions, nb)
		return
	}
	piece := (*allowed_pieces)[fnd]

	// Slide and rotate the piece over the board

	for rot := range 2 {
		if rot == 1 && (piece.Width == piece.Height) {
			// Square pieces are the same no matter the rotation. We
			// can save time by skipping rotations on a square piece.
			continue
		}
		for y := range brd.Height {
			for x := range brd.Width {
				if brd.PlacePiece(piece, x, y, rot == 1) {
					Solve(brd, allowed_pieces, retSolutions)
				}
				brd.RemovePiece(piece)
			}
		}
	}

}
